crates/plugins/mediator-coordination/src/forward/routing.rs

```diff
@@ -0,0 +1,252 @@
+use std::sync::Arc;
+
+use axum::response::{IntoResponse, Response};
+
+use database::Repository;
+use didcomm::{AttachmentData, Message};
+use hyper::StatusCode;
+use mongodb::bson::doc;
+use serde_json::{json, Value};
+
+use crate::{
+    model::stateful::entity::{Connection, RoutedMessage},
+    web::{error::MediationError, AppState, AppStateRepository},
+};
+
+use super::error::RoutingError;
+
+/// Mediator receives forwarded messages, extract the next field in the message body, and the attachments in the message
+/// then stores the attachment with the next field as key for pickup
+pub async fn mediator_forward_process(
+    state: &AppState,
+    payload: Message,
+) -> Result<Message, Response> {
+  
+    let result = handler(state, payload).await.unwrap();
+    Ok(result)
+}
+
+async fn checks(
+    message: &Message,
+    connection_repository: &Arc<dyn Repository<Connection>>,
+) -> Result<String, Response> {
+    let next = message.body.get("next").and_then(Value::as_str);
+    match next {
+        Some(next) => next,
+        None => {
+            let response = (StatusCode::BAD_REQUEST, RoutingError::MalformedBody.json());
+            return Err(response.into_response());
+        }
+    };
+    // Check if the client's did in mediator's keylist
+    let _connection = match connection_repository
+        .find_one_by(doc! {"keylist": doc!{ "$elemMatch": { "$eq": &next}}})
+        .await
+        .map_err(|_| MediationError::RepostitoryError)
+        .unwrap()
+    {
+        Some(connection) => connection,
+        None => {
+            let response = (
+                StatusCode::UNAUTHORIZED,
+                MediationError::UncoordinatedSender.json(),
+            );
+            return Err(response.into_response());
+        }
+    };
+    Ok(next.unwrap().to_string())
```

Comment:
At this stage, `next` should have already been extracted.

A function should perform only one task at a time. If your function is written to retrieve the `next` field of the body, then that should be its sole purpose. Similarly, if it's meant to check a connection, it should only do that. I don't understand why you would separate these from the main function, it is not lengthy.

Reply:
my function is not to retrieve the next field is to make checks and act on the result

Reply:
Your function is extracting the `next` entry of the body and checking for an registered connection.

Reply:
yes the function is called checks

Reply:
and getting next is also a check